### PR TITLE
Update net.py and bgp.py

### LINKED ISSUE #55222: Napalm fails to load (virtualname) in salt/utils/napalm.py due to check on opts too early
### Description of Issue
pip.req has been removed from versions of pip later than 10.
see https://stackoverflow.com/questions/25192794/no-module-named-pip-req
see https://github.com/pypa/pip/pull/4700

### Setup
Came across this running salt-bin whereby build Python3.7.4 with ensurepip, but pip 3 install napalm-base fails since no pip.req.  T

### Steps to Reproduce Issue
Now a clean box install Python 3.7.4 with ensurepip
then just for latest
```
pip3 install --upgrade pip
[root@c664min scripts]# pip3 install napalm-base
Collecting napalm-base
  Using cached https://files.pythonhosted.org/packages/14/2f/e923b0595b8123a58330a95f078fc864c70352d4c74716709a44e793cb8b/napalm-base-1.0.0.tar.gz
    ERROR: Command errored out with exit status 1:
     command: /usr/local/python37/bin/python3.7 -c 'import sys, setuptools, tokenize; sys.argv[0] = '"'"'/tmp/pip-install-umwc_g9i/napalm-base/setup.py'"'"'; __file__='"'"'/tmp/pip-install-umwc_g9i/napalm-base/setup.py'"'"';f=getattr(tokenize, '"'"'open'"'"', open)(__file__);code=f.read().replace('"'"'\r\n'"'"', '"'"'\n'"'"');f.close();exec(compile(code, __file__, '"'"'exec'"'"'))' egg_info --egg-base /tmp/pip-install-umwc_g9i/napalm-base/pip-egg-info
         cwd: /tmp/pip-install-umwc_g9i/napalm-base/
    Complete output (5 lines):
    Traceback (most recent call last):
      File "<string>", line 1, in <module>
      File "/tmp/pip-install-umwc_g9i/napalm-base/setup.py", line 5, in <module>
        from pip.req import parse_requirements
    ModuleNotFoundError: No module named 'pip.req'
    ----------------------------------------
ERROR: Command errored out with exit status 1: python setup.py egg_info Check the logs for full command output.
[root@c664min scripts]# python3.7 --version
Python 3.7.4
[root@c664min scripts]# pip3 --version
pip 19.3.1 from /usr/local/python37/lib/python3.7/site-packages/pip (python 3.7)
[root@c664min scripts]# 
```

Note on Amazon Linux 2 Python 3, has Python 3.7.4 and pip3 v9.0.3 so evenif you upgrade, pip.req still sticks around.  Similarly on RHEL 8, Python 3.6.8 and pip3 v9.0.3.

But this it bites on Debian 10, Python 3.7.3 pip3 v18.1-75

```
root@d10min:/home/david# pip3 install napalm-base
Collecting napalm-base
  Downloading https://files.pythonhosted.org/packages/14/2f/e923b0595b8123a58330a95f078fc864c70352d4c74716709a44e793cb8b/napalm-base-1.0.0.tar.gz
    Complete output from command python setup.py egg_info:
    Traceback (most recent call last):
      File "<string>", line 1, in <module>
      File "/tmp/pip-install-ifpzbzoi/napalm-base/setup.py", line 5, in <module>
        from pip.req import parse_requirements
    ModuleNotFoundError: No module named 'pip.req'
    
    ----------------------------------------
Command "python setup.py egg_info" failed with error code 1 in /tmp/pip-install-ifpzbzoi/napalm-base/
root@d10min:/home/david# 
```

The problem is with older pypi packages which have not allowed for the pip.req removal.
StackOverflow shows a work-around we could use but dislike '_internal' as a solution since does not have a permanence about it.

### Versions Report
All current supported versions

### PR DESCRIPTION
**Pull request by aegiacometti**
### What does this PR do?
1.- Reflect new module name **napalm** instead of **napalm-base**. (runners/net.py and bgp.py)
2.- Catch key-exception for IPv6 if the interface does not have IPv6 enabled (runners/net.py)
3.- Net.find not working because there was an dict_key error at line 228
`AttributeError: 'dict_keys' object has no attribute 'extend'.`
4.- Support for IP secondary address (runners/net.py)
5.- the module was trowing an exemption because a variable list named "compare" was initiated with [None]. After adding items to the list there was a max(compare), and None is not supported for comparing of the list with types None and IPNetwork. So changing the initial value to the worst value as IPNetwork('0.0.0.0/0'), solved the issue.

fixes https://github.com/saltstack/salt/issues/55222

### Previous Behavior
```
sudo salt-run net.find 10.0.12.2
'net' __virtual__ returned False: The napalm-base module could not be imported
```

```
sudo salt-run net.find 10.0.12.2
Details for all interfaces that include network 10.0.12.2/32 - only best match returned
ARP Entries for IP 10.0.12.2
Exception occurred in runner net.find: Traceback (most recent call last):
  File "/usr/lib/python3/dist-packages/salt/client/mixins.py", line 377, in low
    data['return'] = func(*args, **kwargs)
  File "/usr/lib/python3/dist-packages/salt/runners/net.py", line 894, in find
    device, interface, mac = _find_interfaces_mac(ip)
  File "/usr/lib/python3/dist-packages/salt/runners/net.py", line 228, in _find_interfaces_mac
    ip_addresses.extend(interface_ipaddrs.get('ipv6', {}).keys())
AttributeError: 'dict_keys' object has no attribute 'extend'
```

```
sudo salt-run net.find 10.100.12.2

Passed invalid arguments: '>' not supported between instances of 'IPNetwork' and 'NoneType'`
```

### New Behavior
```
sudo salt-run net.find 10.0.12.2
Details for all interfaces that include network 10.0.12.2/32 - only best match returned

    --------------------------------------------------------------------------------------------------------------------------
    |   Device  |    Interface    | Interface Description | IP Addresses | Enabled |  UP  |    MAC Address    | Speed [Mbps] |
    --------------------------------------------------------------------------------------------------------------------------
    | c1_s1_cpe | FastEthernet0/0 |         to_r1         | 10.0.12.2/24 |   True  | True | C2:02:5D:80:00:00 |      10      |
    --------------------------------------------------------------------------------------------------------------------------
ARP Entries for IP 10.0.12.2

    ---------------------------------------------------------------------
    | Age |   Device  |    Interface    |     IP    |        MAC        |
    ---------------------------------------------------------------------
    | 0.0 | c1_s1_cpe | FastEthernet0/0 | 10.0.12.2 | C2:02:5D:80:00:00 |
    ---------------------------------------------------------------------
IP Address 10.0.12.2 is set for interface FastEthernet0/0, on c1_s1_cpe

    --------------------------------------------------------------------------------------------------------------------------
    |   Device  |    Interface    | Interface Description | IP Addresses | Enabled |  UP  |    MAC Address    | Speed [Mbps] |
    --------------------------------------------------------------------------------------------------------------------------
    | c1_s1_cpe | FastEthernet0/0 |         to_r1         | 10.0.12.2/24 |   True  | True | C2:02:5D:80:00:00 |      10      |
    --------------------------------------------------------------------------------------------------------------------------
LLDP Neighbors for interface FastEthernet0/0 on device c1_s1_cpe
None
```

```
sudo salt-run net.find 10.100.12.2
Details for all interfaces that include network 10.100.12.2/32 - only best match returned

    ----------------------------------------------------------------------------------------------------------------------------
    |   Device  |    Interface    | Interface Description |  IP Addresses  | Enabled |  UP  |    MAC Address    | Speed [Mbps] |
    ----------------------------------------------------------------------------------------------------------------------------
    | c1_s1_cpe | FastEthernet0/1 |       to_inside       | 10.100.12.1/24 |   True  | True | C2:02:5D:80:00:01 |      10      |
    ----------------------------------------------------------------------------------------------------------------------------
ARP Entries for IP 10.100.12.2

    ------------------------------------------------------------------------
    | Age  |   Device  |    Interface    |      IP     |        MAC        |
    ------------------------------------------------------------------------
    | 55.0 | c1_s1_cpe | FastEthernet0/1 | 10.100.12.2 | C2:04:5F:8C:00:00 |
    ------------------------------------------------------------------------
IP Address 10.100.12.2 is set for interface FastEthernet0/0, on c1_s1_co-1
LLDP Neighbors for interface FastEthernet0/0 on device c1_s1_co-1
None
```

### CODE REVIEW

File: salt/runners/net.py
Comment (Akm0d) [inline salt/runners/net.py]: A cleaner way to switch to a `set` would be: 
```
ip_addresses = set(interface_ipaddrs.get('ipv4', {}).keys())
ip_addresses.update(set(interface_ipaddrs.get('ipv6', {}).keys()))
```

Comment (aegiacometti) [reply salt/runners/net.py]: Yes, I completely agree. Thanks!

**Review by aegiacometti — COMMENTED** [review]
updated based on comments from Akm0d

**Comment by dwoz** [comment]
@aegiacometti Is it possible to add some tests for these runners?

**Comment by aegiacometti** [comment]
> @aegiacometti Is it possible to add some tests for these runners?
I don't have enough experience to build the test cases from scratch, considering that there aren't any for these modules yet.
After talking with @mirceaulinic about this PR, he commented that my changes are mostly around the imports, and also that this kind of change wouldn't be caught in any unit tests (if there were some already created)

**Comment by aegiacometti** [comment]
@dwoz hi, just following up this PR. Is there anything else I can do?

**Comment by aegiacometti** [comment]
Hi guys! Could you please tell me what to do? Should i wait? Should i do something else? Thanks in advance

**Comment by Akm0d** [comment]
I'll help write some tests for this

**Review by cro — DISMISSED** [review]

**Review by dmurphy18 — DISMISSED** [review]

**Review by Akm0d — DISMISSED** [review]

**Review by Akm0d — DISMISSED** [review]

**Review by Akm0d — APPROVED** [review]

**Review by dwoz — APPROVED** [review]

File: tests/unit/runners/test_bgp.py
Comment (mirceaulinic) [inline tests/unit/runners/test_bgp.py:34]: Before anything, thank you @Akm0d for helping out here. (the following comments are not aiming you or any individual, rather SaltStack's views on this community and contributors)

I'd like however to raise some questions, just for my understanding of how this process works: what does this test? What specific purpose does it serve? If I were to submit a PR now that entirely changes the behaviour of this function, would this test catch it?
I could be wrong, but I'm seeing that it doesn't test anything at all, it barely just confirms the existence of a function. So have we blocked this PR for over 6 months for tests that don't test anything?

Note, I'm not questioning the importance of the tests, but the rule itself: I understand that everything needs a test, I get that; but it's disappointing the way this rule is followed: blindly, robotically, asking contributors to toss in tests when just making a small adjustment (asking to write tests from scratch, when there isn't any) - but everyone's completely happy when throwing some non-sense tests, right? That's just silly. It may make sense to SS as "hey, our code is X% tested", but to contributors like me, it doesn't, it only makes me feel more and more loathe to contribute to this project, I'm disappointed - to put it mildly.

If SS would ever gave a damn about the community as it's lately started trumpeting left and right, would think these decisions thoroughly, also from the perspective of this community, but sadly, it doesn't.

**Comment by Akm0d** [comment]
@mirceaulinic The original change was difficult to test on it's own as it was mostly a change in the imports.  Another part of the difficulty in writing tests for this module is that there were no existing tests.  My goal in writing "tests that don't do anything" was to make it easier to write tests for future changes to the module -- so that future community contributors wouldn't be intimidate by the need to write the entire test suite for this module from scratch to make a small change.   You will also notice that your PR gained significant traction once it was related to a critical issue.  PRs don't get triaged and have their progress tracked the same way as issues.  If an issue is created -- or linked to a PR then it gets through our process much quicker.  We value your contribution so much and appreciate your involvement in the community.  I personally apologize for any frustration you have felt with regards to this and hope we can have even better interactions going forward. @waynew  and I have been [streaming](twitch.tv/saltstackinc) our workflow so that you can see exactly what goes through the SS core team's head as we do things like triage, review PRs, write tests.

File: tests/unit/runners/test_bgp.py
Comment (waynew) [reply tests/unit/runners/test_bgp.py:34]: For what it's worth, we realized that many of our contributors are inexperienced with writing tests, so I've started a Test Clinic, streaming live on Tuesday afternoon and Tuesday mornings (Central/US Time zones).

We have been encouraging contributors to come and learn more about how to write *meaningful* tests, and especially bring their PRs that need tests, and I'm happy to walk through what tests would actually be meaningful, and even start writing tests.

---

When it comes to blindly following this rule, one thing that has stuck out to me is that code missing tests have *significantly* more bugs, in some cases an entire state just wouldn't even work because they're relying on functions or function signatures that literally don't even exist in the module!

And this is code that had been merged into other branches :grimacing:.

Based on this experience, I will take code with blindly added tests that just check that a function even exists, over code without tests, any day. Granted, I'd prefer they have *meaningful* test coverage -- and I'm more than happy to discuss how to achieve that for any PR, during the test clinic, on IRC, Slack, or anywhere else.

---

We absolutely do want meaningful tests, and we realize that's difficult, and are doing as much as we can to support writing tests.

If there's something that you see that we *could* be doing, or doing better, please do bring that to our attention :pray:. Besides the Test Clinic, is there anything we could be doing better to support contributors in writing tests for contributions?

**Pull request merged**
